client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -162,9 +158,6 @@ void handleEndOfSegment(Segment segmentCompleted, boolean fetchSuccesors) throws
         final Map<Segment, List<Long>> segmentToPredecessor;
         if (fetchSuccesors) {
             val successors = getAndHandleExceptions(controller.getSuccessors(segmentCompleted), RuntimeException::new);
-            synchronized (this) {
```

Comment:
As enhancement we could  cache the `delegationToken` in `groupState` since a `delegationToken` is on per Stream level basis. 

The advantages of caching include:
- No additional controller api call to fetch token for every segmentTruncationExcecption received by Reader on its managed segments.
- No additional controller api call to fetch token for every segment acquired by the reader, when a new instance of `SegmentInputStream` is created inside `EventStreamReaderImpl.acquireSegmentsIfNeeded()`.
- Reuse the `delegationToken` returned by `controller.getSuccessors()` api.
(Reposting this comment ...)

Reply:
Makes sense. Created an issue to track this: https://github.com/pravega/pravega/issues/2741.
I do not think we should fix this here.

Reply:
I don't think we should be storing that for security reasons. (See my comment in the issue.)